main.py

```diff
@@ -0,0 +1,338 @@
+    return [d.to_simple() for d in datasetTypes]
+
+
+@subapp.get("/v1/registry/collectionChain/{parent:path}")
+def get_collection_chain(parent: str) -> CollectionSearch:
+    butler = Butler(butler=GLOBAL_BUTLER)
+    chain = butler.registry.getCollectionChain(parent)
+    return chain
+
+
+@subapp.get("/v1/registry/collections")
+def query_collections(regex: Optional[List[str]] = Query(None),
+                      glob: Optional[List[str]] = Query(None),
+                      datasetType: Optional[str] = None,
+                      flattenChains: Optional[bool] = False,
+                      collectionType: Optional[List[str]] = Query(None, regex=COLLECTION_TYPE_RE),
+                      includeChains: Optional[bool] = None) -> List[str]:
+
+    expression_params = ExpressionQueryParameter(regex=regex, glob=glob)
+
+    if collectionType is None:
+        collectionTypes = CollectionType.all()
+    else:
+        # Convert to real collection types
+        collectionTypes = set()
+        for item in collectionType:
+            item = item.upper()
+            try:
+                collectionTypes.add(CollectionType.__members__[item])
+            except KeyError:
+                raise KeyError(f"Collection type of {item} not known to Butler.")
+
+    butler = Butler(butler=GLOBAL_BUTLER)
+    collections = butler.registry.queryCollections(expression=expression_params.expression(),
+                                                   datasetType=datasetType,
+                                                   collectionTypes=collectionTypes,
+                                                   flattenChains=flattenChains,
+                                                   includeChains=includeChains)
+    return list(collections)
+
+
+@subapp.get("/v1/registry/collection/type/{name:path}")
+def get_collection_type(name: str):
+    butler = Butler(butler=GLOBAL_BUTLER)
+    collectionType = butler.registry.getCollectionType(name)
+    return collectionType.name
+
+
+@subapp.put("/v1/registry/collection/{name:path}/{type_}")
+def register_collection(name: str, type_: str, doc: Optional[str] = None) -> str:
+    type_ = type_.upper()
+    try:
+        collectionType = CollectionType.__members__[type_]
+    except KeyError:
+        raise KeyError(f"Collection type of {type_} not known to Butler.")
+
+    butler = Butler(BUTLER_ROOT, writeable=True)
+    butler.registry.registerCollection(name, collectionType, doc)
+
+    return name
+
+
+@subapp.get(
+    "/v1/registry/dataset/{id}",
+    summary="Retrieve this dataset definition.",
+    response_model=Optional[SerializedDatasetRef],
+    response_model_exclude_unset=True,
+    response_model_exclude_defaults=True,
+    response_model_exclude_none=True,
+)
+def get_dataset(id: Union[int, UUID]) -> Optional[SerializedDatasetRef]:
+    butler = Butler(butler=GLOBAL_BUTLER)
+    ref = butler.registry.getDataset(id)
+    if ref is not None:
+        return ref.to_simple()
+    return ref
```

Comment:
What happens if `ref` is `None`? What do you want to return? I suspect you want to raise a 404. An `Optional` return value from an HTTP handler doesn't really make sense.

Reply:
I was happy to return None and assumed that None would be converted to something over the wire that would turn up as a None in the client. I can't remember if I tested that.

Reply:
I don't think JSON can do that. I don't think you can return `null` as a complete JSON document; I thought the top-level element had to be either a list or an object. But maybe I'm wrong?

Reply:
For the record it works absolutely fine. If you return `None`, then on the JSON end you get a `null` and then `response.json()` give you `None`.

Reply:
Oh, huh, TIL. Sounds good then.